File: devmem.go
package main
/*
devman.go
devmem for Omega2 - read/write hardware registers via /dev/mem

Go 1.11.2 32bit downloaded/extracted into $HOME/Programming/go/1.11.2
(not installed, so use export below when at bash prompt)

$export GOROOT=$HOME/Programming/go/1.11.2
$export GOPATH=$HOME/Programming/go/work
$export PATH=$PATH:$GOROOT/bin:$GOPATH/bin/linux_386
$export GOARCH=386
$export GOBIN=$GOPATH/bin/linux_386

bash shell at src folder, cross compile for Omega2-

$GOOS=linux GOARCH=mipsle go build -ldflags="-s -w" devmem.go

then copy to Omega2-
$scp devmem root@Omega_SNAP1:/root
*/

import (
    "fmt"
    "os"
    "syscall"
    "unsafe"
    "strconv"
    "path"
)

const REGBASE = 0x10000000

func usage() {
    pnam := path.Base(os.Args[0])
    fmt.Printf("\n%s r <address>            :read address, decimal\n", pnam)
    fmt.Printf("\n%s h <address>            :read address, hex\n", pnam)
    fmt.Printf("\n%s v <address>            :read address, verbose\n", pnam)
    fmt.Printf("%s w <address> <value>      :write value to address\n", pnam)
    fmt.Printf("%s s <address> <bitmask>    :setbits bitmask at address\n", pnam)
    fmt.Printf("%s c <address> <bitmask>    :clear bits bitmask at address\n", pnam)
    fmt.Printf("\n  <address> = absolute or offset from 0x10000000\n")
    fmt.Printf("  <bitmask> = bits to set or clear\n\n")
    return
}

func memread(mem []byte, addr int, addr_offset int, opt string) {
    v := *(*uint32)(unsafe.Pointer(&mem[addr_offset]))
    v3 := uint8(v>>24); v2 := uint8(v>>16); v1 := uint8(v>>8); v0 := uint8(v)
    if opt == "v" {
        fmt.Printf(
            "\n" +
            "addr: [0x%08X] val: [0x%08X]\n" +
            "\n" +
            "bit 3 3 2 2 2 2 2 2  2 2 2 1 1 1 1 1  1 1 1 1 1 1\n" +
            "pos 1 0 9 8 7 6 5 4  3 2 1 0 9 8 7 6  5 4 3 2 1 0 9 8  7 6 5 4 3 2 1 0\n" +
            "    ...............  ...............  ...............  ...............\n",
            addr, v)
        fmt.Printf(
            "bin ")
        vb := v
        for i := 0; i < 32; i++ {
            b := "0 "
            if (vb & (1<<31)) != 0 { b = "1 " }
            fmt.Printf("%s", b)
            if (i & 7) == 7 { fmt.Printf(" ") }
            vb <<= 1
        }
        fmt.Printf(
            "\n" +
            "hex %11s0x%02X  %11s0x%02X  %11s0x%02X  %11s0x%02X\n" +
            "dec %15d  %15d  %15d  %15d\n\n",
            " ", v3, " ", v2, " ", v1, " ", v0, v3, v2, v1, v0)
    } else if opt == "h"{
        fmt.Printf("0x%08X\n", v)
    } else {
        fmt.Printf("%d\n", v)
    }
}

func main() {
    args := os.Args[1:]
    if len(args) < 2 { usage(); return }
    opt := args[0]
    a, err := strconv.ParseUint(args[1], 0, 32)
    if err != nil { fmt.Printf("invalid address- %s\n", args[1]); return }
    b := uint64(0)
    switch opt {
        case "r":       fallthrough
        case "h":       fallthrough
        case "v":
        case "w":       fallthrough
        case "s":       fallthrough
        case "c":
            if len(args) < 3 { usage(); return }
            b, err = strconv.ParseUint(args[2], 0, 32)
            if err != nil { fmt.Printf("invalid number- %s\n", args[2]); return }
        default: usage(); return
    }

    addr := int(a) &^ 3 | REGBASE   //word aligned, to full address
    val := uint32(b)
    page_addr := addr &^ 0xFFF      //align on page for mmap
    addr_offset := addr - page_addr //offset into mmap page

    f, err := os.OpenFile("/dev/mem", os.O_RDWR|os.O_SYNC, 0666)
    if err != nil { fmt.Printf("failed to open /dev/mem\n"); return }
    defer f.Close()

    mem, err := syscall.Mmap(int(f.Fd()), int64(page_addr), 4096,
        syscall.PROT_READ|syscall.PROT_WRITE, syscall.MAP_SHARED)
    if err != nil { fmt.Printf("failed to mmap /dev/mem\n"); return }

    switch opt {
        case "r":       fallthrough
        case "h":       fallthrough
        case "v":       memread(mem, addr, addr_offset, opt)
        case "w":       *(*uint32)(unsafe.Pointer(&mem[addr_offset])) = uint32(val)
        case "s":       *(*uint32)(unsafe.Pointer(&mem[addr_offset])) |= uint32(val)
        case "c":       *(*uint32)(unsafe.Pointer(&mem[addr_offset])) &^= uint32(val)
    }

    err = syscall.Munmap(mem)
    if err != nil { fmt.Printf("Munmap failed\n") }
}

/*
#230400baud - 40000000/230400 = 173.6
./devmem w 0xd24 3          #UART1_HIGHSPEED = 3
./devmem s 0xd0c 0x80       #UART1_LCR.DLAB = 1
./devmem w 0xd00 1          #UART1_DLL (DLM:DLL = 1)
./devmem w 0xd04 0          #UART1_DLM
./devmem c 0xd0c 0x80       #UART1_LCR.DLAB = 0
./devmem w 0xd28 174        #UART1_SAMPLE_COUNT = 174
./devmem w 0xd2c 87         #UART1_SAMPLE_POINT = 87

#460800baud = 40000000/460800 = 86.8
./devmem w 0xd24 3          #UART1_HIGHSPEED = 3
./devmem s 0xd0c 0x80       #UART1_LCR.DLAB = 1
./devmem w 0xd00 1          #UART1_DLL (DLM:DLL = 1)
./devmem w 0xd04 0          #UART1_DLM
./devmem c 0xd0c 0x80       #UART1_LCR.DLAB = 1
./devmem w 0xd28 87         #UART1_SAMPLE_COUNT = 87
./devmem w 0xd2c 43         #UART1_SAMPLE_POINT = 43

#600000baud = 40000000/600000 = 66.6
./devmem w 0xd24 3          #UART1_HIGHSPEED = 3
./devmem s 0xd0c 0x80       #UART1_LCR.DLAB = 1
./devmem w 0xd00 1          #UART1_DLL (DLM:DLL = 1)
./devmem w 0xd04 0          #UART1_DLM
./devmem c 0xd0c 0x80       #UART1_LCR.DLAB = 1
./devmem w 0xd28 67         #UART1_SAMPLE_COUNT = 67
./devmem w 0xd2c 33         #UART1_SAMPLE_POINT = 33

TODO: be able to run multiple commands
./devman w 0xd34 3 sb 0xd0c 0x80 w 0xd00 1 w 0xd04 0 cb 0xd0c 0x80 w 0xd28 67 w 0xd2c 33
*/

File: devmem.c
//=============================================================================
// devmem
// read/write to hardware registers on Onion Omega2
//=============================================================================

#include <stdio.h>
#include <sys/mman.h>
#include <stdint.h>
#include <fcntl.h>
#include <unistd.h> //close
#include <stdlib.h>
#include <stdbool.h>
#include <libgen.h> //basename

//=============================================================================
// defines
//=============================================================================
#define MAX_ARGS 255                    //limit # of args to program
#define BASE_ADDR 0x10000000            //Omega2 base register address

//=============================================================================
// command buffer struct and global buffer
//=============================================================================
typedef struct {
    char cmd;
    uint32_t addr;
    uint32_t val;
} cmd_buffer_t;
cmd_buffer_t g_cmdbuffer[MAX_ARGS+1];   //store all validated commands/values

//=============================================================================
// show usage
//=============================================================================
static void usage(const char* n){
    printf(
        "\n"
        " %s r <addr>            :read address, return decimal\n"
        " %s h <addr>            :read address, return hex\n"
        " %s b <addr>            :read address, return binary\n"
        " %s v <addr>            :read address, verbose output\n"
        " %s w <addr> <val>      :write address value\n"
        " %s s <addr> <bitmask>  :set bitmask bit(s) at address\n"
        " %s c <addr> <bitmask>  :clear bitmask bit(s) at address\n"
        "\n"
        "    <addr> absolute address or offset address from 0x10000000\n"
        "       address will be automatically word aligned\n"
        "    <val> 32bit value\n"
        "    <bitmask> 32bit bitmask value (1=set or clr specified bit)\n"
        "\n"
        "    numbers can be in any format-\n"
        "       decimal, hex (0x_), binary (0b_), octal (0_)\n"
        "    can provide more than one command at a time, up to a limit\n"
        "       of 255 arguments\n"
        "\n",
        n,n,n,n,n,n,n
    );
    exit(EXIT_FAILURE);
}

//=============================================================================
// print error message, then exit
//=============================================================================
static void error_exit(const char* msg)
{
    if(msg) fprintf(stderr, "%s\n", msg);
    exit(EXIT_FAILURE);
}

//=============================================================================
// value conversion argument error
//=============================================================================
static void arg_error_exit(int argn, const char* str)
{
    fprintf(stderr, "argument# %d -> %s\n", argn, str);
    error_exit("invalid command or value");
}

//=============================================================================
// map physical address to virtual memory, 1 page will be mapped
// error/exit on any failures
// return pointer to virtual address where word is located
//=============================================================================
static uint32_t* regmap(uint32_t addr)
{
    static uint32_t m_page;     //current physical page mapped
    static void* m_vmem;        //current mmap pointer
    static uint32_t m_pagesize; //system page size

    //get system page size (most likely 4096), get/store once
    if(m_pagesize == 0) m_pagesize = sysconf(_SC_PAGE_SIZE);
    //if same page already mapped, no need to map again
    if(m_vmem && (addr / m_pagesize * m_pagesize) == m_page) goto ret;
    //if previously mapped and unmap fails (!0), error
    if(m_vmem && munmap(m_vmem, m_pagesize)){
        error_exit("failed to unmap register memory");
    }

    int fd = open("/dev/mem", O_RDWR);
    if(fd < 0) error_exit("failed to open /dev/mem");
    m_page = addr / m_pagesize * m_pagesize;
	m_vmem = mmap(
        NULL,                   //mmap determines virtual address
        m_pagesize,             //length (just get a whole page)
        PROT_READ | PROT_WRITE, //match file open
        MAP_FILE | MAP_SHARED,
        fd,                     //file
        m_page                  //offset into fd (page aligned)
    );
    close(fd);                  //fd no longer needed
    if(m_vmem == MAP_FAILED) error_exit("failed to mmap register memory");

    ret:
	return &((uint32_t*)m_vmem)[(addr-m_page)/4];
}

//=============================================================================
// convert command line value argument to unsigned integer
// exit on any failure
//=============================================================================
int strto_u32(const char* str, int argn)
{
    if(*str == '-') arg_error_exit(argn, str); //disallow negative numbers
    char *endptr = NULL;
    const char *s = str;
    int base = 0;                   //auto for decimal, hex, octal
    if(s[0] == '0' && s[1] == 'b'){ //check for binary
        base = 2;                   //is binary
        s += 2;                     //skip over "0b"
    }
    uint32_t val = strtoul(s, &endptr, base);
    if(*endptr) arg_error_exit(argn, str); //failed if *endptr not 0
    return val;                     //conversion ok
}

//=============================================================================
// validate everything on command line before doing anything
// (either all good, or all fails, no partital writes)
// store everything in g_cmdbuffer so main loop only has to go through already
//   validated/converted buffer
// g_cmdbuffer[] = { cmd, address, value, ..., 0, 0, 0 }
//=============================================================================
bool validate_command_line(int argc, char** argv)
{
    cmd_buffer_t* cb = &g_cmdbuffer[0];
    for(int i = 0; i < argc; cb++){
        cb->cmd = argv[i][0];
        //only single letter command allowed
        if(argv[i++][1]) arg_error_exit(i, argv[i-1]);
        if(i >= argc) error_exit("incomplete command");
        cb->addr = strto_u32(argv[i], i+1);
        //get addr into 0x1xxxxxxx and word aligned
        cb->addr &= (BASE_ADDR-1);  // 0x0.......
        cb->addr |= BASE_ADDR;      // 0x1.......
        cb->addr &= ~3;             // 0x.......n (n = 0b..00)
        i++;
        switch(cb->cmd){
            case 'r': case 'h': case 'b': case 'v':
                break; //already got address, nothing to do
            case 'w': case 's': case 'c':
                if(i >= argc) error_exit("incomplete command");
                cb->val = strto_u32(argv[i], i+1);
                i++;
                break;
            default:
                arg_error_exit(i, argv[i-1]);
                break;
        }
    }
}

//=============================================================================
// read register 'v' verbose, 'h' hex, 'r' decimal, 'b' binary
//=============================================================================
void regread(char opt, uint32_t addr, uint32_t val)
{
    uint8_t b0 = val;
    uint8_t b1 = val>>8;
    uint8_t b2 = val>>16;
    uint8_t b3 = val>>24;

    if(opt == 'v'){
        printf(
            "\n"
            "addr: [0x%08X] val: [0x%08X]\n"
            "\n"
            "bit 3 3 2 2 2 2 2 2  2 2 2 1 1 1 1 1  1 1 1 1 1 1\n"
            "pos 1 0 9 8 7 6 5 4  3 2 1 0 9 8 7 6  5 4 3 2 1 0 9 8  7 6 5 4 3 2 1 0\n"
            "    ...............  ...............  ...............  ...............\n"
            "bin ",
            addr, val);
        for(int i = 0; i < 32; i++, val<<=1){
            printf("%s", val & (1<<31) ? "1 " : "0 " );
            if((i & 7) == 7) printf(" ");
        }
        printf(
            "\n"
            "hex %11s0x%02X  %11s0x%02X  %11s0x%02X  %11s0x%02X\n"
            "dec %15d  %15d  %15d  %15d\n\n",
            " ", b3, " ", b2, " ", b1, " ", b0, b3, b2, b1, b0);
    } else if(opt == 'h'){
        printf("0x%08X\n", val);
    } else if(opt == 'b'){
        printf("0b ");
        for(int i = 0; i < 32; i++, val<<=1){
            printf("%s", val & (1<<31) ? "1" : "0");
            if((i & 7) == 7) printf(" ");
        }
        printf("\n");
    } else {
        printf("%d\n", val);
    }

}

//=============================================================================
// MAIN
//=============================================================================
int main(int argc, char** argv)
{
    //limit args to a reasonable number
    if(argc > MAX_ARGS) error_exit("argument count exceeded");

    //check argc, if no args show usage
    //let basename modify argv[0]- is ok
    if(argc < 2) usage(basename(argv[0]));

    //validate (starting at first arg)
    validate_command_line(argc-1, &argv[1]);

    //iterate over validated command line buffer
    for(cmd_buffer_t* cb = &g_cmdbuffer[0]; cb->cmd; cb++){
        uint32_t* p32 = regmap(cb->addr);
        switch(cb->cmd){
            case 'r':
            case 'h':
            case 'v':
            case 'b':
                regread( cb->cmd, cb->addr, *p32 );
                break;
            case 'w': *p32 = cb->val;
                break;
            case 's': *p32 |= cb->val;
                break;
            case 'c': *p32 &= ~cb->val;
                break;
        }
    }

    exit(EXIT_SUCCESS);
}
